### PR TITLE
HTTP1 header formatting moved to headers.format_headers and rewritten.

### PR DESCRIPTION
**Pull request by Tronic**
- New implementation is one line of code and twice faster than the old one.
- Whole header block encoded to UTF-8 in one pass.
- No longer supports custom encode method on header values.
- Cookie objects now have `__str__` in addition to `encode`, to work with this.

### CODE REVIEW

**Comment by Tronic** [comment]
Using a small header block to test:
`
%timeit old_impl(h)                                                                                                                                             3.27 µs ± 20.1 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

%timeit new_impl(h)                                                                                                                                          1.63 µs ± 5.61 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
`

**Comment by abuckenheimer** [comment]
you mind sharing the header block / code your using to perf test?

**Comment by Tronic** [comment]
Just something I typed in the interactive prompt. You can easily test it yourself by the IPython `%timeit` magic command.

**Comment by Tronic** [comment]
Benchmarking this branch:
````
$ hey -z 30s http://127.0.0.1:8000/

Summary:
  Total:	30.0018 secs
  Slowest:	0.0451 secs
  Fastest:	0.0002 secs
  Average:	0.0018 secs
  Requests/sec:	28217.1343
  
  Total data:	11005332 bytes
  Size/request:	13 bytes

Response time histogram:
  0.000 [1]	|
  0.005 [811481]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  0.009 [30926]	|■■
  0.014 [3251]	|
  0.018 [732]	|
  0.023 [117]	|
  0.027 [19]	|
  0.032 [30]	|
  0.036 [0]	|
  0.041 [0]	|
  0.045 [7]	|


Latency distribution:
  10% in 0.0006 secs
  25% in 0.0009 secs
  50% in 0.0014 secs
  75% in 0.0021 secs
  90% in 0.0033 secs
  95% in 0.0044 secs
  99% in 0.0074 secs

Details (average, fastest, slowest):
  DNS+dialup:	0.0000 secs, 0.0002 secs, 0.0451 secs
  DNS-lookup:	0.0000 secs, 0.0000 secs, 0.0000 secs
  req write:	0.0000 secs, 0.0000 secs, 0.0019 secs
  resp wait:	0.0017 secs, 0.0002 secs, 0.0450 secs
  resp read:	0.0001 secs, 0.0000 secs, 0.0053 secs

Status code distribution:
  [200]	846564 responses
````

git master 1e4b1c4d1a907af559dcf91b697debd60d94bbd7:
````
$ hey -z 30s http://127.0.0.1:8000/

Summary:
  Total:	30.0037 secs
  Slowest:	0.0415 secs
  Fastest:	0.0002 secs
  Average:	0.0018 secs
  Requests/sec:	28169.2870
  
  Total data:	10987392 bytes
  Size/request:	13 bytes

Response time histogram:
  0.000 [1]	|
  0.004 [798935]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  0.008 [39669]	|■■
  0.013 [5071]	|
  0.017 [1040]	|
  0.021 [295]	|
  0.025 [115]	|
  0.029 [32]	|
  0.033 [18]	|
  0.037 [0]	|
  0.041 [8]	|


Latency distribution:
  10% in 0.0007 secs
  25% in 0.0009 secs
  50% in 0.0013 secs
  75% in 0.0021 secs
  90% in 0.0033 secs
  95% in 0.0045 secs
  99% in 0.0079 secs

Details (average, fastest, slowest):
  DNS+dialup:	0.0000 secs, 0.0002 secs, 0.0415 secs
  DNS-lookup:	0.0000 secs, 0.0000 secs, 0.0000 secs
  req write:	0.0000 secs, 0.0000 secs, 0.0017 secs
  resp wait:	0.0017 secs, 0.0002 secs, 0.0412 secs
  resp read:	0.0001 secs, 0.0000 secs, 0.0064 secs

Status code distribution:
  [200]	845184 responses
````

In other words, no difference.

**Comment by Tronic** [comment]
Benchmarking the `encode_body` alternatives. The old implementation uses try-except to handle data that are not string, and the except part takes a lot of time. It turns out that f-strings enable us to handle any str-convertible types very fast, with no penalty on values that already are strings.

```text
                123 to bytes                             "foo" to bytes
try-except      ███████████████████████████ 928 ns        202 ns ██████
f-string        ████████ 280 ns                            197 ns █████
always str()    ███████████ 387 ns                      313 ns ████████
```

File: sanic/headers.py [hunk 1]
```diff
@@ -165,3 +168,30 @@ def parse_host(host: str) -> Tuple[Optional[str], Optional[int]]:
...
+def format_http1(headers: HeaderIterable) -> bytes:
+    """Convert a headers iterable into HTTP/1 header format.
+
+    - Outputs UTF-8 bytes where each header line ends with \\r\\n.
+    - Values are converted into strings if necessary.
+    """
+    return "".join(f"{name}: {val}\r\n" for name, val in headers).encode()
+
+
+def format_http1_response(
```
Comment (abuckenheimer) [inline sanic/headers.py]: I think this may belong in `sanic/response.py`

File: sanic/response.py [hunk 2]
```diff
@@ -168,11 +135,8 @@ def __init__(
         self._cookies = None
 
     def output(self, version="1.1", keep_alive=False, keep_alive_timeout=None):
-        # This is all returned in a kind-of funky way
-        # We tried to make this as fast as possible in pure python
-        timeout_header = b""
-        if keep_alive and keep_alive_timeout is not None:
-            timeout_header = b"Keep-Alive: %d\r\n" % keep_alive_timeout
+        if "Content-Type" not in self.headers:
+            self.headers["Content-Type"] = self.content_type
```
Comment (abuckenheimer) [inline sanic/response.py]: I think this means we can delete 148->150

**Comment by Tronic** [comment]
@abuckenheimer re your benchmarks -- I broke passing int to HTTPResponse in my another PR, so you'll have to change that (provided that the change is accepted) -- you need to pass str or bytes (or something alike but not integers).

**Comment by Tronic** [comment]
Will look at that conflict later.

**Comment by Tronic** [comment]
Requesting review.

**Comment by ahopkins** [comment]
@huge-success/sanic-release-managers

**Pull request merged**